docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,492 @@
+* **Upgrade stage**: Used to upgrade from one EdgeX release to another.
+* **Run stage**: The stage where EdgeX services are running in production.
+
+In particular, it is necessary to forestall entry into the "run stage"
+until either the install stage or the upgrade stage has completed.
+There is currently no "uninstall stage" defined.
+
+The stage-gate mechanism must work in the following environments:
+
+* docker-compose in Linux on a single node
+* docker-compose in Microsoft* Windows* on a single node
+* docker-compose in Apple* MacOS* on a single node
+
+#### Stage-gate design
+
+Docker-compose v2.x used to have a `depends_on / condition` directive
+that would test a service's HEALTHCHECK
+and block startup
+until the service was "healthy".
+Unfortunately, this feature was removed in 3.x docker-compose.
+(This feature was also unsuppported in swarm mode as well.)
+
+The current prevailing mechanism
+for blocking service startup in Docker environments
+is to wait for open TCP ports.
+There are several tools for this, including
+[wait-for-it](https://github.com/vishnubob/wait-for-it),
+[dockerize](https://github.com/jwilder/dockerize), and
+[wait-for](https://github.com/Eficode/wait-for).
+The TCP mechanism does not depend on any underlying OS features
+such as System V IPC,
+can work in a multi-node installation,
+and is portable across platforms.
+
+The core of the stage-gate design is
+to add special services to the docker-compose scripts
+that implement a simple "200 OK" HTTP HEALTHCHECK
+service after the required initialization has been completed.
+The service will respond "200 OK" to all requests,
+will terminate on SIGTERM,
+and the listening port will not be opened
+until the required processing has finished.
+
+The following special services will be added to the framework:
+
+* `install` will initialize a fresh EdgeX installation.
+* `upgrade_*release*_to_*release*_` will be a special container
+  that is either a no-op,
+  or initiate an upgrade from an *n-1* release to the *n* release of EdgeX.
+* `secretstore_ready` will bring Vault online and issue fresh access tokens for all services.
+* `persistence_ready` will wait for the system databases to be started.
+* `ready_to_run` will wait for all of the above.
+
+With a few exceptions,
+the majority of EdgeX services will block on `ready_to_run`'s HEALTHCHECK.
+
+### Seeding of initial secrets
+
+For "mass production" of EdgeX appliances,
+it is not feasible to manually touch and configure each appliance.
+It is desirable to inject certain secrets directly into the bootstrapping flow.
+These secrets include:
+
+* TBD
+* TBD
+* TBD
+
+After secret store initialization,
+a JSON file containing seed secrets will be processed and injected into the KV secrets engine.
+This JSON file will be of the format:
+
+<pre>
+{
+  "path/to/secret": {
+    anything
+  }, ...
+}
+</pre>
+
+This file will be fetched from a preconfigured URL (file://, https://) passed to the bootstrapping container.
```

Comment:
How about `(File or HTTPS)` rather than `(file://, https://)` ? current make it look like you left something out and require too muck thinking to realize these are placeholders... ;-)

Reply:
fair enough.